File: java-to-kotlin/src/main/java/travelator/Legs.java
package travelator;

import java.time.Duration;
import java.util.List;
import java.util.Optional;

public class Legs {

    public static Optional<Leg> findLongestLegOver(
            List<Leg> legs,
            Duration duration
    ) {
        Leg result = null;
        for (Leg leg: legs) {
            if (isLonerThan(leg, duration)) {
                if (result == null || isLonerThan(leg, result.getPlannedDuration())) {
                    result = leg;
                }
            }
        }
        return Optional.ofNullable(result);

    }

    private static boolean isLonerThan(Leg leg, Duration duration) {
        return leg.getPlannedDuration().compareTo(duration) > 0;
    }

}

File: java-to-kotlin/src/main/kotlin/travelator/Legs.kt
package travelator

import java.time.Duration

object Legs {

    fun longestLegOver(
        legs: List<Leg>,
        duration: Duration
    ): Legs? {
        var result: Leg? = null

        for (leg in legs) {
            if (leg.isLongerThan(duration))
                if (result == null || leg.isLongerThan(result.plannedDuration))
                    result = leg
        }
        return result
    }

    fun longestLegOver(
        legs: List<Leg>,
        duration: Duration
    ): Leg? {
        val longestLeg: Leg? = legs.maxByOrNull { Leg::getPlannedDuration }
        return if (longestLeg != null && longestLeg.plannedDuration > durattion) {
            longestLeg
        } else {
            null
        }
    }




    private fun Leg.isLongerThan(duration: Duration) = plannedDuration.compareTo(duration) > 0



}